internal/lib/channel.ml

```diff
@@ -0,0 +1,39 @@
+(****************************************************************************)
+(*                           the diy toolsuite                              *)
+(*                                                                          *)
+(* Jade Alglave, University College London, UK.                             *)
+(* Luc Maranget, INRIA Paris-Rocquencourt, France.                          *)
+(*                                                                          *)
+(* Copyright 2010-present Institut National de Recherche en Informatique et *)
+(* en Automatique and the authors. All rights reserved.                     *)
+(*                                                                          *)
+(* This software is governed by the CeCILL-B license under French law and   *)
+(* abiding by the rules of distribution of free software. You can use,      *)
+(* modify and/ or redistribute the software under the terms of the CeCILL-B *)
+(* license as circulated by CEA, CNRS and INRIA at the following URL        *)
+(* "http://www.cecill.info". We also give a copy in LICENSE.txt.            *)
+(****************************************************************************)
+
+(** Utilities for handling in_channel and out_channel. *)
+
+(* in_channel utilities *)
+
+let rec iter_lines f chan =
+  try begin
+    (f (input_line chan)) ; iter_lines f chan
+  end
+  with End_of_file -> ()
+
+let map_lines f chan =
+  let ret = ref [] in
+  iter_lines (fun l -> ret := !ret @ [ f l ]) chan ;
+  !ret
```

Comment:
This code is quadratic (construction a list by repetitive append at the end).
Better to construct the list by `ret := f l :: !ret` and return `List.rev !ret`.

Reply:
Done.